### PR TITLE
[#29] Add `fromFoldableWithIndex`

### LINKED ISSUE #29: Add `fromFoldableWithIndex`
I've seen it mentioned a [couple](https://discord.com/channels/864614189094928394/865617619464749081/887326007105814548) [times](https://discord.com/channels/864614189094928394/867149806178664478/1032345440248668242) in the Discord that `Object.fromFoldableWithIndex` could be a useful addition, although maybe there's a reason this doesn't exist yet that I'm overlooking?

### PR DESCRIPTION
**Pull request by pete-murphy**
**Description of the change**

Add `fromFoldableWithIndex` and tests. Fixes #29.

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

File: test/Test/Foreign/Object.purs [hunk 1]
```diff
@@ -150,6 +153,14 @@ objectTests = do
     quickCheck (O.lookup "1" nums == Just "one"  <?> "invalid lookup - 1")
     quickCheck (O.lookup "2" nums == Nothing     <?> "invalid lookup - 2")
 
+  log "fromFoldableWithIndex & key collision"
+  do
+    let numsMap = M.fromFoldable [Tuple 0 "zero", Tuple 1 "what", Tuple 1 "one"]
+        nums = O.fromFoldableWithIndex show numsMap
+    quickCheck (O.lookup "0" nums == Just "zero" <?> "invalid lookup - 0")
+    quickCheck (O.lookup "1" nums == Just "one"  <?> "invalid lookup - 1")
+    quickCheck (O.lookup "2" nums == Nothing     <?> "invalid lookup - 2")
```
Comment (pete-murphy) [inline test/Test/Foreign/Object.purs]: Oops, this isn't actually testing key collision on the object, since the `M.fromFoldable` handles that. I'll update.

File: src/Foreign/Object.purs [hunk 2]
```diff
@@ -221,6 +222,14 @@ fromFoldable l = runST do
   ST.foreach (A.fromFoldable l) \(Tuple k v) -> void $ OST.poke k v s
   pure s
 
+-- | Create an `Object a` from an indexed foldable collection, using the
+-- | specified function for converting the index to `String`
```
Comment (JordanMartinez) [inline src/Foreign/Object.purs]: This should also state something about key collisions using the later values.

File: src/Foreign/Object.purs [hunk 3]
```diff
@@ -221,6 +222,15 @@ fromFoldable l = runST do
...
 
+-- | Create an `Object a` from an indexed foldable collection, using the
+-- | specified function for converting the index to a `String` key. On key
+-- | collisions, the last value for the duplicate key will be kept.
+fromFoldableWithIndex :: forall f k v. FoldableWithIndex k f => (k -> String) -> f v -> Object v
+fromFoldableWithIndex f l = runST do
+  s <- OST.new
+  forWithIndex_ l \k v -> OST.poke (f k) v s
+  pure s
+
```
Comment (JordanMartinez) [inline src/Foreign/Object.purs:231]: Thoughts on adding a variant `fromFoldableWithIndex'` that assumes the `k` is already a `String`?

**Review by JordanMartinez — APPROVED** [review]
LGTM otherwise.

File: src/Foreign/Object.purs (hunk 3 above)
Comment (pete-murphy) [reply src/Foreign/Object.purs:231]: I'm now thinking that makes sense as the default API, based on the two use cases[^1] cited in the issue. They both say that `Object.fromFoldableWithIndex` would be useful specifically in the case of `FoldableWithIndex k f ~ Map String`.
```haskell
fromFoldableWithIndex :: forall f a. FoldableWithIndex String f => f a -> Object a
```
I think non-`String` keys would probably be the less common case, and in that case you might as well _also_ throw in a function for combining values on conflict (instead of defaulting to overriding earlier values, which could be undesirable)
```haskell
fromFoldableWithIndexWith :: forall f k v. FoldableWithIndex k f => (k -> String) -> (v -> v -> v) -> f v -> Object v
```
basically support the (probably) common case with `fromFoldableWithIndex` and then give more granular control with `fromFoldableWithIndexWith` (not sure about that naming 😄 )

[^1]: https://discord.com/channels/864614189094928394/865617619464749081/887326007105814548
      https://discord.com/channels/864614189094928394/867149806178664478/1032345440248668242

Comment (garyb) [reply src/Foreign/Object.purs:231]: I think we can skip `fromFoldableWithIndexWith` because you could `mapKeys` (or equivalent) the input first.

It would be more efficient to convert while doing this, but you could say the same about accepting mapping functions in `fromFoldable`s in general if we wanted to start going down that route.

Comment (pete-murphy) [reply src/Foreign/Object.purs:231]: I suppose
```haskell
fromFoldableWithIndexWith
  :: forall f k a b
   . FoldableWithIndex k f
  => (k -> a -> Tuple String b)
  -> (b -> b -> b)
  -> f a
  -> Object b
```
would be the most general variation, but I don't know if that's too complicated.

Comment (pete-murphy) [reply src/Foreign/Object.purs:231]: > It would be more efficient to convert while doing this, but you could say the same about accepting mapping functions in `fromFoldable`s in general if we wanted to start going down that route.

Didn't see your comment til after I posted mine, that makes sense to skip `fromFoldableWithIndexWith`. So would just adding
```haskell
fromFoldableWithIndex :: forall f a. FoldableWithIndex String f => f a -> Object a
```
make sense?

**Review by garyb — APPROVED** [review]
I think we keep it simple for now :+1:

Can add extra variants if we find we need them later!
